flatsurf/geometry/hyperbolic.py

```diff
@@ -0,0 +1,4521 @@
+        base ring, we cannot produce defining half spaces since these would
+        require equations over a quadratic extension as well::
+
+            sage: H.half_circle(0, 2).start()._half_spaces()
+            Traceback (most recent call last):
+            ...
+            ValueError: square root of 32 not a rational number
+
+        Over an inexact ring, the number of half spaces might not be minimal
+        since we cannot decide the finiteness of a point::
+
+            sage: H = HyperbolicPlane(RR)
+
+            sage: H(I)._half_spaces()
+            [{(x^2 + y^2) + 2.00000000000000*x - 1.00000000000000 ≤ 0},
+             {x ≥ 0},
+             {(x^2 + y^2) - 1.00000000000000 ≥ 0}]
+
+            sage: H(0)._half_spaces()
+            [{(x^2 + y^2) + x ≤ 0}, {x ≥ 0}, {(x^2 + y^2) ≥ 0}]
+
+        """
+        x0, y0 = self.coordinates(model="klein")
+
+        if not self.parent().base_ring().is_exact() or self.is_finite():
+            return HyperbolicHalfSpace._merge_sorted(
+                # x ≥ x0
+                [self.parent().half_space(-x0, 1, 0, model="klein")],
+                # y ≥ y0
+                [self.parent().half_space(-y0, 0, 1, model="klein")],
+                # x + y ≤ x0 + y0
+                [self.parent().half_space(x0 + y0, -1, -1, model="klein")])
+        else:
+            return HyperbolicHalfSpace._merge_sorted(
+                # left of the line from (0, 0) to this point
+                [self.parent().half_space(0, -y0, x0, model="klein")],
+                # right of a line to this point with a starting point right of (0, 0)
+                [self.parent().half_space(-x0*x0 - y0*y0, y0 + x0, y0 - x0, model="klein")])
+
+    def is_finite(self):
+        if not self.parent().base_ring().is_exact():
+            raise NotImplementedError("cannot decide finiteness over inexact rings")
+
+        if isinstance(self._coordinates, HyperbolicOrientedGeodesic):
+            return False
+
+        x, y = self.coordinates(model="klein")
+        return x*x + y*y < 1
+
+    def coordinates(self, model="half_plane", ring=None):
+        r"""
+        Return coordinates of this point in ``ring``.
+
+        If ``model`` is ``"half_plane"``, return projective coordinates in the
+        Poincaré half plane model.
+
+        If ``model`` is ``"klein"``, return Euclidean coordinates in the Klein model.
+
+        If no ``ring`` has been specified, an appropriate extension of the base
+        ring of the :class:`HyperbolicPlane` is chosen where these coordinates
+        live.
+        """
+        # TODO: Implement ring.
+        # TODO: Fix documentation.
+
+        if model == "half_plane":
+            x, y = self.coordinates(model="klein")
+
+            if x == 0 and y == 1:
+                raise ValueError(f"{self} has no coordinates in the upper half plane")
+
+            denominator = 1 - y
+
+            # TODO: Do something better here
```

Comment:
This bit is very annoying : I have non-ideal points that suddenly became ideals... The example is not crazy, it is just a regular octagon with floating point coordinates on which I apply a small matrix.

Reply:
Yes, this is not necessary anymore. I'll push a fix.

Reply:
The fix is in. It's not great (there is still a hack in there.) But it's hopefully a bit more reasonable now.